wgpu-core/src/adapter.rs

```diff
@@ -0,0 +1,246 @@
+use crate::{
+    backend,
+    hub::{GfxBackend, Global, IdentityFilter, Token},
+    id::AdapterId,
+    Backend,
+};
+
+#[cfg(feature = "serde")]
+use serde::{Deserialize, Serialize};
+
+pub use hal::adapter::AdapterInfo;
+use hal::Instance;
+
+#[derive(Debug)]
+pub struct RawAdapter<B: hal::Backend> {
+    pub(crate) raw: hal::adapter::Adapter<B>,
+}
+
+#[allow(clippy::large_enum_variant)]
+#[derive(Debug)]
+pub(crate) enum Adapter {
```

Comment:
It seems to me that there is a lot of logic in this change that doesn't need to be.
This enum, in particular, shouldn't be needed. Instead, we have different "hubs" containing arrays of `hal::adapter::Adapter<B>` (each for a different `B`), and then we figure out which array to access based on the ID, which contains the backend enum it in (see `gfx_select!` implementation).

Is there a way we can write this with more relying on the existing infrastructure and less repeating of all the per-backend configuration flags?

Reply:
Well yes, it's just to refactor this, I have to be able to pass the adapters between functions somehow.

I can remove `Adapter`, but then I will just make a struct that will collect these hubs instead. Personally I thought not calculating IDs made the code cleaner, specifically lines like these:
[`if selected < adapters_vk.len() {`](https://github.com/gfx-rs/wgpu/pull/478/files#diff-5bc89207c08401c5ad6c2186d12c149dL296)
[`selected -= adapters_vk.len();`](https://github.com/gfx-rs/wgpu/pull/478/files#diff-5bc89207c08401c5ad6c2186d12c149dL308)

Give the word and I will return it as it was 👍!

Reply:
"I have to be able to pass the adapters between functions somehow." - can't you pass the `AdapterId` between functions?
"that will collect these hubs instead" - not clear what you mean by this?

Reply:
Apologies for the lack of clarity!
> "I have to be able to pass the adapters between functions somehow." - can't you pass the `AdapterId` between functions?

The new function `enumerate_adapters` shares functionality between `pick_adapter`. Specifically `inputs.find(Backend::{BACKEND})` and `instance.{BACKEND}.enumerate_adapters()`.
Most of the work here is mainly just abstracting this away into a function, which is now [`Adapter::enumerate`](https://github.com/gfx-rs/wgpu/pull/478/files#diff-fa745ea317d41e5354c904b8cc61cb3cR36-R87).

So after abstracting it, I have to somehow return these two things to use them inside `enumerate_adapters` and `pick_adapter`. This is why I created `Adapter`.
Of course I can just return a tuple with all four "hubs" instead, which comes to your next question.

> "but then I will just make a struct that will collect these hubs instead" - not clear what you mean by this?

Instead of returning a tuple with four elements, I would just create a struct to make the code a little cleaner, this is just a minor idea I had, irrelevant, I could also just destructure the tuple.


I did all of this mainly to be able to abstract as much as possible away to avoid code duplication.
Also `Adapter::register`, which again will be complicated to abstract away if I didn't have this `Adapter` enum.